source/mir/cv/imgproc/imfilter.d

```diff
@@ -0,0 +1,469 @@
+                (input, temp, hm, vm, output);
+            break;
+        default:
+            inner_filtering_impl!(Non_SIMD!T, apply_horizontal_kernel!T, apply_vertical_kernel!T)
+                (input, temp, hm, vm, output);
+        }
+    }
+
+    if (border_handling != BORDER_REPLICATE)
+    {
+        return ERROR_INVALID_BORDER_METHOD;
+    }
+
+    borders_replicate_impl(output, ksize);
+
+    return 0;
+}
+
+pragma(inline, true)
+void inner_filtering_impl
+(
+    alias InstructionSet,
+    alias hkernel,
+    alias vkernel
+)
+(
+    Slice!(Contiguous, [2], InstructionSet.Scalar*) input,
+    Slice!(Contiguous, [2], InstructionSet.Scalar*) temp,
+    Slice!(Contiguous, [1], InstructionSet.Scalar*) hmask,
+    Slice!(Contiguous, [1], InstructionSet.Scalar*) vmask,
+    Slice!(Contiguous, [2], InstructionSet.Scalar*) output
+)
+in
+{
+    assert(input.shape == temp.shape, "Incompatible input and temporary buffer slice");
+    assert(temp.shape == output.shape, "Incompatible input and output buffer slice");
+}
+body
+{
+    alias T = InstructionSet.Scalar;
+    alias V = InstructionSet.Vector;
+    immutable velems = InstructionSet.elementCount;
+    immutable tbytes = T.sizeof;
+    immutable vbytes = V.sizeof;
+    immutable ksize = hmask.length;
+    immutable rows = input.length!0;
+    immutable cols = input.length!1;
+
+    static if (Is_SIMD!InstructionSet)
+    {
+        // It this is SIMD instructions set, allocate vectors for kernels
+
+        auto hk = cast(V[])alignedAllocate(ksize * vbytes, 16); // horizontal simd kernel
+        auto vk = cast(V[])alignedAllocate(ksize * vbytes, 16); // vertical simd kernel
+
+        scope (exit)
+        {
+            deallocate(hk);
+            deallocate(vk);
+        }
+
+        foreach (i; 0 .. ksize)
+        {
+            hk[i].array[] = hmask[i];
+        }
+
+        foreach (i; 0 .. ksize)
+        {
+            vk[i].array[] = vmask[i];
+        }
+    }
+    else
+    {
+        // ... otherwise just use input buffers.
+        auto hk = hmask._iterator[0 .. ksize];
+        auto vk = vmask._iterator[0 .. ksize];
+    }
+
+    // Get pointers to where vectors are loaded (has to be public)
+    static auto toPtr(ref T e)
+    {
+        return &e;
+    }
+
+    auto _in = input.universal.map!toPtr;
+    auto _tmp = temp.universal.map!toPtr;
+    auto _out = output.universal.map!toPtr;
+
+    // Stride buffers to match vector indexing.
+    auto a = _in[0 .. $ - ksize, 0 .. $ - ksize].strided!1(velems);
+    auto t = _tmp[0 .. $ - ksize, 0 .. $ - ksize].strided!1(velems);
+    auto b = _out[0 .. $ - ksize, 0 .. $ - ksize].strided!1(velems);
+
+    // L1 cache block tiling (under 8192 bytes)
+    immutable rtiling = 32;
+    immutable ctiling = max(size_t(1), 256 / (tbytes * ksize));
+
+    // Process blocks
+    zip!true(a, t, b)
+    .blocks(rtiling, ctiling)
+  //.parallel
+    .each!((b) {
```

Comment:
please foreach loops instead of `each`. THe are more readable. `each` is good when the function is already defined.